lib/IRGen/GenReflection.cpp

```diff
@@ -1006,6 +1006,65 @@ void IRGenModule::emitBuiltinTypeMetadataRecord(CanType builtinType) {
   builder.emit();
 }
 
+class MultiPayloadEnumDescriptorBuilder : public ReflectionMetadataBuilder {
+  CanType type;
+  const FixedTypeInfo *ti;
+
+public:
+  MultiPayloadEnumDescriptorBuilder(IRGenModule &IGM,
+                                    const NominalTypeDecl *nominalDecl)
+    : ReflectionMetadataBuilder(IGM) {
+    type = nominalDecl->getDeclaredType()->getCanonicalType();
+    ti = &cast<FixedTypeInfo>(IGM.getTypeInfoForUnlowered(
+        nominalDecl->getDeclaredTypeInContext()->getCanonicalType()));
+  }
+
+  void layout() override {
+    // TODO: Don't try to emit MPE descriptors for unbound generic enums
+    // Related: getEnumImplStrategy() below asserts that `type` is not an unbound generic
+    // if (type is generic) {
+    //   return;
+    // }
+
+    addTypeRef(type, CanGenericSignature());
+
+    auto &strategy = getEnumImplStrategy(IGM, type);
+    bool isMPE = strategy.getElementsWithPayload().size() > 1;
+    assert(isMPE && "");
+
+    const TypeInfo &TI = strategy.getTypeInfo();
+    auto bits1 = strategy.getTagBitsForPayloads();
+    auto bits2 = strategy.getBitMaskForNoPayloadElements();
+    // Look at EnumImplStrategy::emitGetEnumTag to see how it works
```

Comment:
TODO:  Get the spare bitmask from the right place.  `bits1` and `bits2` above look like either one of them might be right.

Reply:
Looking at [GenEnum.cpp:5572](https://github.com/apple/swift/blob/a193d2d791f9ae972c1689318f3dc20cbc7be751/lib/IRGen/GenEnum.cpp#L5572), it really looks like `getTagBitsForPayloads()` should provide exactly what I need here.  But for an MPE where every payload is a class reference, it is giving me 0xe000000000000000 instead of the 0xff00000000000007 that I expect.  @slavapestov What am I missing here?

Reply:
Looks like you might have an enum in your test case that requires 3bits to represent the tag? `getTagBitsForPayloads()` returns the bits that are used for tagging. Not the spare bits available.

```
    // The common spare bits actually used for a tag in the payload area.       
    SpareBitVector PayloadTagBits;  
```

`IRGenModule::getHeapObjectSpareBits` would tell you what spare bits we use for heap objects.

Assuming you are looking for the common spare bits among the payloads of an multi payload enum (?) I think you want (to add) a function that returns:

```
    // The spare bits shared by all payloads, if any.                           
    // Invariant: The size of the bit vector is the size of the payload in bits,
    // rounded up to a byte boundary.                                           
    SpareBitVector CommonSpareBits;     
```

Reply:
After a few hours of tracing code, I found [`applyFixedSpareBitsMask`](https://github.com/apple/swift/blob/a193d2d791f9ae972c1689318f3dc20cbc7be751/lib/IRGen/FixedTypeInfo.h#L219), which seems to do exactly that.

Reply:
I am not sure what mask you are looking for. 

But I don't believe that `multipayloadEnum->applyFixedSpareBitsMask() == CommonSpareBits`. I would expect `multipayloadEnum->applyFixedSpareBitsMask()` to return the remaining spare bits after subtracting the bit used for the tag (also see the code at https://github.com/apple/swift/blob/877fb2b84110d88c11c39e98bd9b3ac292f9c284/lib/IRGen/GenEnum.cpp#L6642)

A useful tools debugging inspecting this is to use the command:

```
swiftc -Xllvm -debug-only=enum-layout fileWithEnums.swift -c
```

This will enable the debug printing code at https://github.com/apple/swift/blob/877fb2b84110d88c11c39e98bd9b3ac292f9c284/lib/IRGen/GenEnum.cpp#L6901

If you add code to print the `CommonSpareBits` you can view how they relate to the other values. On first appearance I would think that you can reconstruct CommonSpareBits by combining `applyFixedSpareBitsMask` and `getTagBitsForPayloads` (assuming we don't use extra tag bits at least?).

Reply:
The right code here turns out to be the following, which computes the common spare bits for all payloads (which is precisely what the runtime is looking for):
```
      for (auto enumCase : strategy.getElementsWithPayload()) {
        cast<FixedTypeInfo>(enumCase.ti)->applyFixedSpareBitsMask(IGM, spareBits);
      }
```